### PR TITLE
Multipart Decoder

### LINKED ISSUE #503: Multipart Requests

### PR DESCRIPTION
**Pull request by smehta91**

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/experiment/Part.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package zhttp.experiment
+
+import io.netty.buffer.ByteBufUtil
+import io.netty.handler.codec.http.multipart.{Attribute => JAttribute, FileUpload, HttpData, InterfaceHttpData}
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/experiment/Part.scala]: no `J` prefix required.

File: zio-http/src/main/scala/zhttp/experiment/ContentDecoder.scala [hunk 2]
```diff
@@ -1,36 +1,81 @@
...
+      override def offer(a: HttpContent): UIO[Unit] = q.offer(a).unit
+      override def poll: UIO[List[Int]]             = q.map(_.content().array().length).takeAll
+    }
+  }
 
+  object Multipart0 {}
   case object Text extends ContentDecoder[Any, Nothing, Any, String]
 
   case class Step[R, E, S, A, B](state: S, next: (A, S, Boolean) => ZIO[R, E, (Option[B], S)])
       extends ContentDecoder[R, E, A, B]
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/experiment/ContentDecoder.scala]: case class Stream[R, E, S, A, B](state: S, next: (A, S, Boolean) => ZIO[R, E, (Option[B], S)])
      extends ContentDecoder[R, E, A, Stream[B]]

File: zio-http/src/main/scala/zhttp/experiment/Part.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import zio.Chunk
+
+sealed trait Part
+
+object Part {
+  def fromHTTPData(data: multipart.InterfaceHttpData): Part = data match {
+    case data: multipart.HttpData =>
+      data match {
+        case attribute: multipart.Attribute =>
+          val b     = attribute.content()
```
Comment (smehta91) [inline zio-http/src/main/scala/zhttp/experiment/Part.scala:14]: rename variables.

File: zio-http/src/main/scala/zhttp/http/ContentDecoder.scala [hunk 4]
```diff
@@ -15,31 +18,102 @@ object ContentDecoder {
   case class Step[R, E, S, A, B](state: S, next: (A, S, Boolean) => ZIO[R, E, (Option[B], S)])
       extends ContentDecoder[R, E, A, B]
 
-  private[zhttp] case class BackPressure[B](queue: Option[Queue[B]] = None, isFirst: Boolean = true) {
+  private[zhttp] case class BackPressure[B](acc: Option[B] = None, isFirst: Boolean = true) {
```
Comment (smehta91) [inline zio-http/src/main/scala/zhttp/http/ContentDecoder.scala:21]: Think of better name

File: zio-http/src/main/scala/zhttp/http/ContentDecoder.scala [hunk 5]
```diff
@@ -15,31 +18,102 @@ object ContentDecoder {
...
+  def toJRequest(req: Request): HttpRequest = new DefaultHttpRequest(
+    HttpVersion.HTTP_1_1,
+    req.method.asHttpMethod,
+    req.url.asString,
+    Header.disassemble(req.headers),
+  ) // todo: remove this
+
+  def multipartDecoder(req: Request): Task[PostBodyDecoder[Throwable, HttpContent, Part]] =
+    Task(new PostBodyDecoder[Throwable, HttpContent, Part] {
+      private val decoder = new HttpPostRequestDecoder(new DefaultHttpDataFactory(false), toJRequest(req)) //
```
Comment (smehta91) [inline zio-http/src/main/scala/zhttp/http/ContentDecoder.scala:43]: remove unwanted comment

**Comment by smehta91** [comment]
Closing in favor of https://github.com/dream11/zio-http/pull/515

**Pull request closed without merging**